hedy.py

```diff
@@ -876,36 +881,47 @@ def all_print_arguments(input_string, level, lang='en'):
 
     return AllPrintArguments(level).transform(program_root)
 
-
-@v_args(meta=True)
+@v_args(meta=True)  # Note that setting meta=True here is required regardless of the annotation of the Filter class
 class IsValid(Filter):
     # all rules are valid except for the "Invalid" production rule
     # this function is used to generate more informative error messages
     # tree is transformed to a node of [Bool, args, command number]
-    def program(self, args, meta):
+
+    def program(self, args, meta=None):
         if len(args) == 0:
             return False, InvalidInfo("empty program")
-        return super().program(args)
+        return super().program(args, meta)
 
     def error_invalid_space(self, args, meta):
         # return space to indicate that line starts in a space
-        return False, InvalidInfo(" ", line=meta.line, column=meta.column)
+        return False, InvalidInfo(" ", line=args[0][2].line, column=args[0][2].column), meta
 
     def error_print_nq(self, args, meta):
         # return error source to indicate what went wrong
-        return False, InvalidInfo("print without quotes", line=meta.line, column=meta.column)
+
+        # Boryana Jan 22
+        # The meta variable passed as a parameter here contains the metadata of the current node.
+        # The args parameter is an array of the triples, of which the last part holds the metadata of the
+        # IMMEDIATE children. Two important points:
+        # - If an error rule has more than 2 levels of nested children, we lose the metadata of the
+        # deeper ones and get only the data of the immediate children. This does work for now, because the parse trees
+        # of all error grammar rules are only 1 level deep. For now.
+        # - At level 5, the error_print_nq can have multiple arguments and we cannot assume that it is the first one
+        # that has a missing quote. Further analysis is required to pinpoint where the quote should be added.
```

Comment:
@boryanagoncharenko First of all, thank you of course!

I get the problem, so if there are error rules with deeper levels, we only have meta available from the children and not the children's children. So we want to have the meta from the deeper nodes also available (in `args` again?).

But like you said, our error rules are currently only 2 levels. So I am not sure if we can implement this already, without have those deeper level error rules.

Reply:
Well there are 2 issues here:

* error grammar rules can be deeper -> yes, you are right we could leave that for now, although I would prefer this to be fixed fo the future
* error_print_nq can have multiple arguments -> further analysis is needed to pinpoint the right location of the missing quote

Can youat least look into point 2 @dolfein? This is also visible from the tests around this that do not always show the right place.

Reply:
Let's consider the following program at level 5:
```
name is 'B'
L is 1, 2, 3
print 'Person ' name ' said ' L at random things'
```
The print node in the AST produced by the above code will have 5 arguments: 
- (text, "Person '")
- (var, ['name'])
- (text, "' said '")
- (list_access, [...])
- (text, "things'")

We want to skip the children of types var and list access and examine only the ones of type 'text'. So when we reach the error_print_nq method in the IsValid transformer, we need the type, value and meta information for each children. However, IsValid is a transformer which by default is transforming bottom up. This means that by the time we reach the error_print_nq, its children are transformed, so we lose the information we need.

The easiest way to address the problem at hand is to use a visitor instead. Consider the following code:
```
class IsValidVisitor(Visitor):
    def error_print_nq(self, tree):
        pass
...
def is_program_valid(program_root, input_string, level, lang):
    visitor = IsValidVisitor()
    visitor.visit(program_root)
...
```
[
NOTE Please have in mind that there is a tiny improvement/fix in the grammar that needs to be done in the grammar of level 5. The line 
```error_print_no_quotes: _PRINT _SPACE (textwithoutspaces | list_access | var) (_SPACE (textwithoutspaces | list_access | var))*  -> error_print_nq``` needs to change to ```error_print_no_quotes: _PRINT _SPACE (var | textwithoutspaces | list_access) (_SPACE (var | textwithoutspaces | list_access))*  -> error_print_nq``` instead. The reason is that var is more restrictive than textwithoutspaces and, therefore, needs to appear first.
]

In the above suggestion when we enter the error_print_nq we have the whole tree, including the type, values and meta information of the children. Figuring out which child has the missing quote is simpler now.
 
However, I am not sure whether changing the current implementation of the IsValid/Filter classes is something we want. When using a visitor we will fail at the first occurrence of an error while at the moment the transformer allows us to accumulate multiple errors. @Felienne maybe you could confirm this was the original idea behind the implementation? If so, we could make a hybrid transformer which returns the tree and accumulates the collected errors but this might increase the complexity.

Also before making any changes to the structure of the Filter, we need to ensure every code path is covered with tests.

Reply:
> The print node in the AST produced by the above code will have 5 arguments:
> 
> - (text, "'Person '")
> - (var, ["name"])
> - (text, "' said '")
> - (list_access, [...])
> - (text, "things'")

I am getting these arguments, am I doing something wrong? This is before IsValid.
<img width="426" alt="image" src="https://user-images.githubusercontent.com/71043277/150903591-3a8fc3cb-53b5-449b-87f3-7b3ca25236d2.png">

I also came across these 2 issues: https://github.com/lark-parser/lark/issues/252 and https://github.com/lark-parser/lark/issues/556. Can we solve this with an Interpreter?

Reply:
>1. From this PR I assume we want to use the meta information to provide the concrete place of the error, e.g. a missing quote.

Yes, meta provides us with the line and (first) column of a **node**. If we know in which node the error is, we can look within that node for the exact column and store that in InvalidInfo.column.

>2. If 1) is correct, please consider the following program. Should we add or remove a quote? Both options yield perfectly valid programs.
>```
>name is Hedy
>print 'name
>```

I would say, in this case, remove the quote. I think this one is more interesting:
```
name is Hedy
print 'my name is name
```
>3. If 1) is correct, please consider the following program. Should we add a quote and where? Or should we remove a quote and which one?
>```
>people is mom, dad, Emma, Sophie
>print 'today the dishes are done by ' people at random'
>```

Removing the last quote would be optimal.

>4. Do you need meta information for other IsValid violations, e.g. 'invalid command' or 'invalid space'. If so, can you provide examples of program inputs (with their Hedy level) and the expected meta information?

Invalid space is, according to the comment, always a leading space, so meta.column is unnecessary here.
Invalid command can be useful in the future. I tried this example but a wrong error is raised. But the input should hopefully still give you an idea, we want to know the column of prind.
<img width="496" alt="image" src="https://user-images.githubusercontent.com/71043277/151077094-5f58e416-e2f2-4fa9-9479-2dced47cfd5a.png">

But I remember that @Felienne also said that meta could be useful for other stuff as well.

Reply:
> > 2. If 1) is correct, please consider the following program. Should we add or remove a quote? Both options yield perfectly valid programs.
> > 
> > ```
> > name is Hedy
> > print 'name
> > ```
> 
> I would say, in this case, remove the quote. 

Reading this, I think maybe we are all confused about what the others are thinking. So let me clarify where I want to go with program repair. It is *not* about finding the exact right location of an error. We only want to have the location to have a general sense of where to start trying to remove, add or swap out characters.

Absent of the right meta information, we can still very much do program repair, but then we just would do totally random things and see where they get us. The meta info helps us be more effective.

So there is not the situation that "remove the quote" is the right answer here. With program repair, we will generate many solutions, maybe one for adding *and* one for removing the quote, and we somehow evaluate the results. Maybe we present all solutions that parse to the user and heave them pick, but that is of later concern. The sky is the limit in terms of tries, but the bottom line is that we do not need a precise or perfect location for meta, just a good enough one.

> Invalid space is, according to the comment, always a leading space, so meta.column is unnecessary here. Invalid command can be useful in the future. I tried this example but a wrong error is raised. But the input should hopefully still give you an idea, we want to know the column of prind. <img alt="image" width="496" src="https://user-images.githubusercontent.com/71043277/151077094-5f58e416-e2f2-4fa9-9479-2dced47cfd5a.png">

Please @dolfein, for now focus on only the quotes and maybe indeed the leading spaces, that is tricky enough for your project. Invalid command could first use some other improvements (from me)

> But I remember that @Felienne also said that meta could be useful for other stuff as well.

Yes, but let's not consider that now, we need to scope your project down zo it becomes doable.

I propose the following:
* Based on the meta information there is now, you start to work on code that does random mutation (with the code you already wrote) and evaluates mutants (mutants = programs changed by code)
* In the mean time, @boryanagoncharenko will see if she can improve the meta information

Reply:
@Felienne thank you for the clarification. This sounds like a nice project and I am looking forward to its results.

@dolfein I highly recommend keeping the IsValid transformer as it is for now. Please create your own traverser and plug it right before the IsValid.transform() invocation and handle the error cases that are interesting to you. Make sure that you are raising the needed exception immediately in the visitor (as opposed to the IsValid transformer which returns information about the error and an exception is raised later). Once you are ready with your project, I will clean up the IsValid transformer so that it does not repeat what you have already covered.

You can use the following PR as a starting point: https://github.com/Felienne/hedy/pull/1815. It defines a new Visitor and plugs it right before the IsValid.transform() method. Once you are in the Visitor, you will get a tree object which has a meta property and all its children will have a meta property too. This solves the problem of getting the meta of the immediate children, since you now get the full unprocessed tree annotated with meta for every node. If you need the original program code, you can use self.code. So this is all the information you need. Note that the UnquotedTextException is thrown in the visitor. 

Let me know if you have any questions about the meta information.